cdap-master/src/main/java/co/cask/cdap/data/tools/Main.java

```diff
@@ -158,115 +162,28 @@ private Action parseAction(String action) {
   }
 
   private void performUpgrade(Injector injector) throws Exception {
-    // Upgrade system datasets
-    upgradeSystemDatasets(injector);
+    LOG.debug("Performing setup for Upgrade ...");
+    Upgrade upgradeSetup = new UpgradeSetup();
+    upgradeSetup.upgrade(injector);
 
-    // Upgrade all user hbase tables
-    upgradeUserTables(injector);
+    LOG.debug("Upgrading System and User Datasets ...");
+    Upgrade dsUpgrade = new DatasetUpgrade();
```

Comment:
as mentioned before, don't pass in the injector.  These classes should take the required objects in from their constructors, and the constructors should be private and annotated with ```@Inject```.  Then here you use the injector to get DatasetUpgrade, MDSUpgrade, etc